File: netdb.i2p2/fixedapp.py
#
# Required software:
# sqlite3 3.6.23.1 or newer
#    http://www.sqlite.org/
# unixodbc
#    http://www.unixodbc.org/
# pyodbc-3.0.6 or newer
#    http://www.ch-werner.de/sqliteodbc/
# werkzeug
#    http://werkzeug.pocoo.org/
# python 2.6 or 2.7
#    http://python.org/
#


# Modify as needed, or use a symlink.
netdbdir = 'netdb'
database = 'Driver=SQLite;DATABASE=I2PnetDb'


from werkzeug import BaseRequest, BaseResponse, ETagResponseMixin, escape, SharedDataMiddleware
from werkzeug.exceptions import HTTPException
import os
from time import time, gmtime
import calendar
#import time
from random import choice
import pyodbc


#
# Not needed?
#
#import sha


# db format
#
#  table client ipaddress, time of access
#
#  table html index, time of creation


#
# Init database if it does not exist.
#

def checkdatabase():
    cnxn = pyodbc.connect(database)
    cur = cnxn.cursor()

    try:
        cur.execute("select * from client")
    except pyodbc.Error:
        #print ("Creating new table 'client'")
        cur.execute("create table client (ip string, whn float)")
        cnxn.commit()
    try:
        cur.execute("select * from entry")
    except pyodbc.Error:
        #print ("Creating new table 'entry'")
        cur.execute("create table entry (whn float, wht string)")
        cnxn.commit()
    cnxn.close()




#    try:
#        cur.execute("select * from html")
#    except pyodbc.Error:
#        cur.execute("create table client (index, time, html)")


#    cur.execute("insert into foo values(1, 'Me')")
#    cnxn.commit()
#    cur.execute("select * from foo")
#    cnxn.close()

class Request(BaseRequest):
    """Useful subclass of the default request that knows how to build urls."""

    def __init__(self, environ):
        BaseRequest.__init__(self, environ)


class Response(BaseResponse, ETagResponseMixin):
    """Subclass of base response that has a default mimetype of text/html."""
    default_mimetype = 'text/html'

def application(environ, start_response):
    """The WSGI application that connects all together."""
    checkdatabase()
    req = Request(environ)
    remote = req.remote_addr
    now = time()
    old = now - 3600
    oldest = now - 7200
    t = gmtime(now)
    nowtag = calendar.timegm((t.tm_year, t.tm_mon, t.tm_mday, t.tm_hour, 0, 0, t.tm_wday, t.tm_yday, t.tm_isdst))

    cnxn = pyodbc.connect(database)
    cur = cnxn.cursor()
    #
    # Prune database from peers that are over 1 hour old
    #
    cur.execute("delete from client where whn < ?", old)
    cnxn.commit()
    #
    # Prune database from entries that are over 2 hours old
    #
    cur.execute("delete from entry where whn < ?", oldest)
    cnxn.commit()

    #
    # Get peer info
    #
    cur.execute("select * from client where ip = ?", remote)
    info = cur.fetchall()
    # page
    path = req.path[1:]


    if path == '':
        page = u'<html><head><title>NetDB</title></head><body><ul>%s</ul></body></html>'

        if len(info) == 0:
            # tag the ip as new
            cur.execute("insert into client values (?, ?)", (remote, now))
            # see if we have a list already, and use that
            cur.execute("select * from entry where whn = ?", nowtag)
            stuff = cur.fetchall()
            if (len(stuff) == 0):
                # generate links
                entries = os.listdir(netdbdir)
                new = []
                if len(entries) > 150:
                    # select some randomly
                    for i in xrange(100):
                        while True:
                            sel = choice(entries)
                            if not sel.startswith('routerInfo-'):
                                continue
                            if sel not in new:
                                new.append(sel)
                                cur.execute("insert into entry values (?, ?)", (nowtag, sel))
                                break
                else:
                    for sel in entries:
                        if not sel.startswith('routerInfo-'):
                            continue
                        if sel not in new:
                            new.append(sel)
                            cur.execute("insert into entry values (?, ?)", (nowtag, sel))
                entries = new
            else:
                # Use what we already generated
                entries = []
                for junk,sel in stuff:
                    entries.append(sel)
        else:
            # use old list based on date in database, i.e. sends the same as before.
            junk, last = info[0]
            t = gmtime(last)
            oldtag = calendar.timegm((t.tm_year, t.tm_mon, t.tm_mday, t.tm_hour, 0, 0, t.tm_wday, t.tm_yday, t.tm_isdst))
            cur.execute("select * from entry where whn = ?", oldtag)
            stuff = cur.fetchall()
            entries = []
            for junk,sel in stuff:
                entries.append(sel)

        res = ''
        for entry in entries:
            # Already sanitized above.
            #if not entry.startswith('routerInfo-'):
            #    continue
            res += '<li><a href="%s">%s</a></li>' % (entry, entry)
        resp = Response(page % res, mimetype='text/html')
        resp.add_etag()

    elif path == 'robots.txt':
        dat = u"User-agent: *\nDisallow: /routerInfo-*.dat$\n"
        resp = Response(dat, mimetype='text/plain')
        resp.add_etag()

    else:
        if len(info) == 0:
            # 404 This is to prevent cross seed site scrapes
            resp = Response("Hi! Having fun?", status=403, mimetype='text/plain')
        else:
            #
            # zap badness from path
            #
            #print "path given is " + repr(path)

            chkpath = path.replace("'","").replace("/","")
            # check to see if we have this entry in the database at all.
            cur.execute("select * from entry where wht = ?",  chkpath)
            chk = cur.fetchall()
            if len(chk) == 0:
                resp = Response("Do you not have anything better to do?", status=400, mimetype='text/plain')
            else:
                try:
                    # load file
                    f = open(os.path.join(netdbdir, path), 'rb')
                    resp = Response(f.read(), mimetype='application/octet-stream')
                    f.close()
                    resp.add_etag()
                except IOError:
                    # 404
                    resp = Response("Owch! Could not find file! It got deleted before you could get a copy, sorry.", status=404, mimetype='text/plain')
    cnxn.commit()
    cnxn.close()
    return resp(environ, start_response)

if __name__ == '__main__':
    from werkzeug import run_simple
    run_simple('localhost', 5007, application)

File: netdb.i2p2/fixedapp.php
 <?php
# Reseed script
#

define('DATABASE', '/var/www/reseed.db');
define('NUM_ROUTERS', 50);
define('NETDB_DIR', '/home/i2p/.i2p/netDb');
define('DEBUG', false);

/*
-------------------------------------
Add following to .htaccess:
Options +FollowSymlinks
RewriteEngine On
RewriteRule ^(.*\.dat)$ /index.php?file=$1 [L]
RewriteCond %{REQUEST_FILENAME} !index.php
RewriteCond %{REQUEST_FILENAME} !robots.txt

-------------------------------------

Do not touch if you don't know what you're doing.
*/

# Database handle
class ReseedDatabase extends SQLite3
{
    protected $db;
    protected $createStatement1 = 'create table client (ip string, whn int)';
    protected $createStatement2 = 'create table entry (whn int, wht string)';

    public function __construct($file, $flags = SQLITE3_OPEN_READONLY)
    {
        parent::__construct($file, $flags);
    $this->busyTimeout(10000);
        // this only has to happen once, no sense calling it on every request
        //$this->_create();
    }

    public function insertEntry($time, $entry)
    {
        $results = $this->exec(sprintf('insert into entry (whn,wht) values (%d,"%s")', $time, $entry));
        if ($results == true) return true;
        return false;
    }

    public function insertClient($time, $host)
    {
        $results = $this->exec(sprintf('insert into client (ip,whn) values ("%s",%d)', $host, $time));
        if ($results == true) return true;
        return false;
    }

    public function getEntries($time)
    {
        $results = $this->query(sprintf('select * from entry where whn = "%s"', $time));
        if ($results instanceof SQLite3Result) {
            $t = array();
            while ($j = $results->fetchArray(SQLITE3_ASSOC)) {
                $t[] = $j;
            }
            return $t;
        }
        return false;
    }

    public function getEntriesFromClient($host)
    {
        $results = $this->query(sprintf('select e.wht from entry e, client c where c.whn=e.whn and c.ip = "%s"', $host));
        if ($results instanceof SQLite3Result) {
            $t = array();
            while ($j = $results->fetchArray(SQLITE3_ASSOC)) {
                $t[] = $j['wht'];
            }
            return $t;
        }
        return false;
    }

    public function getClient($host)
    {
        $results = $this->query(sprintf('select * from client where ip = "%s"', $host));
        if ($results instanceof SQLite3Result) return $results->fetchArray();
        return false;
    }

    public function cleanup()
    {
        $this->exec(sprintf('delete from client where whn < %d', (time() - 3600)));
        $this->exec(sprintf('delete from entry where whn < %d', (time() - 7200)));
    }

    private function _create()
    {
        if ($this->query('select * from entry') == false && $this->query('select * from client') == false) {
            $this->exec($this->createStatement1);
            $this->exec($this->createStatement2);
        }
    }

}


$user_agent = $_SERVER['HTTP_USER_AGENT'];
$remote_ip = $_SERVER["REMOTE_ADDR"];

if (true !== DEBUG && !strstr($user_agent, 'Wget/1.11.4')) {
    header('HTTP/1.0 400 Access Denied');
    header('Content-Type: text/plain');
    die("Access denied.");
}

// if they're requesting a file, let's try to send it to them
if (isset($_GET['file'])) {
    // sanitize the input filename
    $file = htmlentities($_GET['file'], ENT_QUOTES, 'UTF-8', false);
    $file = str_replace('..', '', str_replace('/', '', $file));
    $file = realpath(NETDB_DIR . DIRECTORY_SEPARATOR . $file);
    $filename = basename($file);

    // make sure they are requesting a real file
    if (!$file || empty($filename) ||
        NETDB_DIR !== substr($file, 0, strlen(NETDB_DIR)) ||
        preg_match("/^routerInfo-(.+)\=.dat$/", $filename) !== 1 ||
        !file_exists($file) ||
        is_dir($file)
    ) {
        header('HTTP/1.0 404 Not Found');
        header('Content-Type: text/plain');
        die('Not found.');
    }

    $db = new ReseedDatabase(DATABASE, SQLITE3_OPEN_READONLY);
    $res = $db->getEntriesFromClient($remote_ip);
    $db->close();
    unset($db);
    if ($res === false || !in_array($filename, $res)) {
        header('HTTP/1.0 404 Not Found');
        header('Content-Type: text/plain');
        die('Not found.');
    }

    header('Content-Type: application/octet-stream');
    header('Content-Transfer-Encoding: binary');
    header('Content-Length: ' . filesize($file));
    header(sprintf('Content-Disposition: attachment; filename="%s"', $filename));

    ob_clean();
    flush();
    readfile($file);

    die();
}

$page = '<html><head><title>NetDB</title></head><body><ul>%s</ul></body></html>';
$entry = '<li><a href="%s">%s</a></li>';

$db = new ReseedDatabase(DATABASE, SQLITE3_OPEN_READONLY);
if ($db->getClient($remote_ip) === false) {
    $time = time();

    $db->close();
    unset($db);
    $db = new ReseedDatabase(DATABASE, SQLITE3_OPEN_READWRITE);
    $db->insertClient($time, $remote_ip);
    $db->close();
    unset($db);

    # generate new list
    $clientRouterList = array();
    if (is_dir(NETDB_DIR) && is_readable(NETDB_DIR)) {
        $clientRouterList = array_map('basename', glob(NETDB_DIR . DIRECTORY_SEPARATOR . 'routerInfo-*=.dat', GLOB_ERR));
    }

    $db = new ReseedDatabase(DATABASE, SQLITE3_OPEN_READWRITE);
    foreach (array_rand($clientRouterList, NUM_ROUTERS) as $router) {
        $db->insertEntry($time, $clientRouterList[$router]);
    }
} else {
    // run cleanup 50% of requests
    if (rand(0, 100) < 50) {
       $db->close();
       unset($db);
       $db = new ReseedDatabase(DATABASE, SQLITE3_OPEN_READWRITE);
       $db->cleanup();
    }
}

// to release locks
$db->close();
unset($db);
$db = new ReseedDatabase(DATABASE, SQLITE3_OPEN_READONLY);

# use old list based on date in database, i.e. sends the same as before
$clientRouterList = $db->getEntriesFromClient($remote_ip);

$db->close();
unset($db);

header('Content-Type: text/html');
die(sprintf($page, implode('', array_map(function($single) use ($entry) { return sprintf($entry, $single, $single);}, $clientRouterList))));

